File: services/busyness-index/vikunja_client.py
"""Minimal read-only Vikunja REST API client.

Credentials come from environment variables VIKUNJA_URL / VIKUNJA_TOKEN
unless passed explicitly (useful for tests).
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Optional

import requests


@dataclass
class VikunjaProject:
    id: int
    title: str

    @classmethod
    def from_api(cls, data: dict) -> "VikunjaProject":
        return cls(id=data["id"], title=data.get("title", ""))


@dataclass
class VikunjaTask:
    id: int
    title: str
    done: bool
    project_id: int
    due_date: Optional[str]
    start_date: Optional[str]
    created: Optional[str]
    updated: Optional[str]
    done_at: Optional[str]
    priority: Optional[int]

    @classmethod
    def from_api(cls, data: dict) -> "VikunjaTask":
        return cls(
            id=data["id"],
            title=data.get("title", ""),
            done=bool(data.get("done", False)),
            project_id=data.get("project_id"),
            due_date=data.get("due_date"),
            start_date=data.get("start_date"),
            created=data.get("created"),
            updated=data.get("updated"),
            done_at=data.get("done_at"),
            priority=data.get("priority"),
        )


class VikunjaClient:
    def __init__(self, base_url: Optional[str] = None, token: Optional[str] = None, timeout: float = 10.0):
        self.base_url = (base_url or os.environ["VIKUNJA_URL"]).rstrip("/")
        self.token = token or os.environ["VIKUNJA_TOKEN"]
        self.timeout = timeout
        self._session = requests.Session()
        self._session.headers.update({"Authorization": f"Bearer {self.token}"})

    def list_projects(self) -> list[VikunjaProject]:
        r = self._session.get(f"{self.base_url}/api/v1/projects", timeout=self.timeout)
        r.raise_for_status()
        return [VikunjaProject.from_api(p) for p in r.json()]

    def list_tasks_in_project(self, project_id: int, per_page: int = 50) -> list[VikunjaTask]:
        tasks: list[VikunjaTask] = []
        page = 1
        while True:
            r = self._session.get(
                f"{self.base_url}/api/v1/projects/{project_id}/tasks",
                params={"page": page, "per_page": per_page},
                timeout=self.timeout,
            )
            r.raise_for_status()
            batch = r.json()
            if not batch:
                break
            tasks.extend(VikunjaTask.from_api(t) for t in batch)
            if len(batch) < per_page:
                break
            page += 1
        return tasks

    def list_all_tasks(self) -> list[VikunjaTask]:
        tasks: list[VikunjaTask] = []
        for project in self.list_projects():
            tasks.extend(self.list_tasks_in_project(project.id))
        return tasks

File: artifacts/api-server/src/lib/vikunjaClient.ts
// Minimal, read-only wrapper around Vikunja's REST API (v1). Vikunja is a
// third-party app (not something we control the source of), so unlike
// pf-cwh/FitnessForge there's no "add a /api/public/summary route" option —
// Aiportal has to call Vikunja's own API directly with a dedicated token.

export interface VikunjaProject {
  id: number;
  title: string;
}

export interface VikunjaTask {
  id: number;
  title: string;
  done: boolean;
  due_date: string;
  start_date: string;
  project_id: number;
}

// Vikunja (Go zero-value time) represents "no date set" as year 1, e.g.
// "0001-01-01T00:00:00Z" — see tasktracker/README.md's gantt-sorting notes.
function isUnsetDate(iso: string | undefined): boolean {
  return !iso || iso.startsWith("0001-");
}

export class VikunjaClient {
  constructor(
    private readonly baseUrl: string,
    private readonly token: string,
  ) {}

  private async request<T>(path: string): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      headers: { Authorization: `Bearer ${this.token}` },
      signal: AbortSignal.timeout(8000),
    });
    if (!res.ok) throw new Error(`Vikunja API ${path} -> HTTP ${res.status}`);
    return res.json() as Promise<T>;
  }

  listProjects(): Promise<VikunjaProject[]> {
    return this.request<VikunjaProject[]>("/api/v1/projects");
  }

  async listTasksInProject(projectId: number): Promise<VikunjaTask[]> {
    const all: VikunjaTask[] = [];
    for (let page = 1; ; page++) {
      const pageTasks = await this.request<VikunjaTask[]>(
        `/api/v1/projects/${projectId}/tasks?page=${page}&per_page=50`,
      );
      if (!pageTasks || pageTasks.length === 0) break;
      all.push(...pageTasks);
      if (pageTasks.length < 50) break;
    }
    return all;
  }
}

export interface UpcomingTask {
  id: number;
  title: string;
  projectId: number;
  projectTitle: string;
  dueDate: string;
  startDate: string | null;
  overdue: boolean;
}

export interface VikunjaSummary {
  windowDays: number;
  tasks: UpcomingTask[];
  totalTasks: number;
  doneTasks: number;
  completionRate: number | null;
  upcomingCount: number;
  overdueCount: number;
  busyIndex: number | null;
}

// Tasks due within `windowDays` (including already-overdue ones), across
// every project, incomplete only. `tasks` is capped for display; the counts
// below are computed from the FULL fetched set, not the capped list.
export async function fetchVikunjaSummary(
  baseUrl: string,
  token: string,
  windowDays = 7,
  maxTasks = 8,
): Promise<VikunjaSummary> {
  const client = new VikunjaClient(baseUrl, token);
  const projects = await client.listProjects();
  const projectTitleById = new Map(projects.map((p) => [p.id, p.title]));

  const tasksByProject = await Promise.all(
    projects.map((p) => client.listTasksInProject(p.id)),
  );
  const allTasks = tasksByProject.flat();

  const totalTasks = allTasks.length;
  const doneTasks = allTasks.filter((t) => t.done).length;
  const completionRate = totalTasks > 0 ? Math.round((doneTasks / totalTasks) * 100) : null;

  const now = Date.now();
  const horizon = now + windowDays * 24 * 60 * 60 * 1000;

  const upcomingAll: UpcomingTask[] = allTasks
    .filter((t) => !t.done && !isUnsetDate(t.due_date))
    .map((t) => ({
      id: t.id,
      title: t.title,
      projectId: t.project_id,
      projectTitle: projectTitleById.get(t.project_id) ?? "",
      dueDate: t.due_date,
      startDate: isUnsetDate(t.start_date) ? null : t.start_date,
      overdue: new Date(t.due_date).getTime() < now,
    }))
    .filter((t) => new Date(t.dueDate).getTime() <= horizon)
    .sort((a, b) => new Date(a.dueDate).getTime() - new Date(b.dueDate).getTime());

  const overdueCount = upcomingAll.filter((t) => t.overdue).length;
  const upcomingCount = upcomingAll.length - overdueCount;

  // 忙碌指數：逾期任務、待辦任務量、完成率三項加權，逾期權重最高（最直接的
  // 壓力訊號），完成率用「落後程度」（100-completionRate）而非直接用完成率，
  // 分數越高代表越忙／壓力越大，封頂 100。這是第一版粗略公式，數字本身沒有
  // 精確的科學意義，用來抓「大概是不是在超載」的相對趨勢。
  const behindRate = completionRate !== null ? 100 - completionRate : 50;
  const busyIndex = Math.min(100, Math.round(overdueCount * 15 + upcomingCount * 5 + behindRate * 0.3));

  return {
    windowDays,
    tasks: upcomingAll.slice(0, maxTasks),
    totalTasks,
    doneTasks,
    completionRate,
    upcomingCount,
    overdueCount,
    busyIndex,
  };
}
